Hash.7.pod
=head1 The Real Life Systems Container library - Hashes

by Ronald Landheer-Cieslak <ronald@landheer.com>, Feb 2004
(updated from Sept. 2002)

=head2 What are hashes?

Any Perl programmer will tell you that a hash is an I<associative array>: any 
I<value> in it is I<associated> with a I<key>. The wonderful thing about hashes 
are that they are very fast: because of the way hashing works, you can find a 
specific value in a has much faster than you would be able to in an array. 
Another very nice thing is that you can use anything you want as a key to your 
hash, while for an array, you can only use an integer value as an index.

=head2 How does hashing work?

Internally, a hash is an array of mappings, in which the keys are mapped to a 
value. Both are C<void*> pointers, hence the ability to use anything as a 
key/value. To know where to put the value in the hash, a calculation is done 
with the key which results in the "estimated" position of the value in the 
hash. The system then looks whether that position is still empty and, if so, 
puts the key and the value at that position in the hash. If the position was 
not empty, the system looks for an empty one "next to" the mapped entry (this 
is called I<linear hashing>). If the key was already used in the hash, the new 
value replaces the old one.

The hashing-system takes care of memory allocation itself, so you never have 
to worry about how much memory you are using for the hash. However, whenever 
you ask for listed information about a hash (like the list of keys that the 
hash contains) memory is allocated that will never be used for the hash itself, 
and that you are responsible for freeing. As I<a priori> you don't have to know 
how big the buffer is that you'll get, there are no functions that let you 
allocate the buffer yourself (if you want one, feel free to contribute it).

=head2 Why make a hashing system in C++ if mapping already exists?

Though mapping already exists in the Standard Template Library, there are three 
simple reasons why I made the Hash class:

=over 4

=item 1

because there is no such thing as mapping in C

=item 2

because I needed the exercise: it's been a while since I last did any C++ 
programming, so I wanted to freshen up on my know-how

=item 3

because STL mappings don't allow you to save a map (and all the associated 
information) to a file - the library's hashes do

=back

=head2 How to use these classes

The library provides you with five classes, which you may find more or less 
useful: the C<Hash> class, which is the base class of all other classes, the 
C<StringHash> class, which is a hash for which you can use a string as a key; 
the C<NocaseStringHash>, which is the same as StringHash, but is not 
case-sensitive; the C<IniHash> class, which is designed to read, interprete 
and write .ini configuration files; and the C<IntHash> class, with which you 
can use an integer as key.

=begin text

    +-Figure 1----------------+
    |  Hash                   |
    |  /  \                   |
    |  |  StringHash          |
    |  |    \                 |
    |  |    NocaseStringHash  |
    |  |      \               |
    |  |      IniHash         |
    |  |                      |
    |  IntHash                |
    +-------------------------+
      
=end text

Though C++ allows you to initialise a base class with one of its descendants, 
except for StringHash and NocaseStringHash (which you can initialise with 
StringHash as well as vice versa) this is not a good idea: the way the hashes 
work internally depends completely on what they expect their keys to be, which 
basically means you should not expect the base class to understand the keys of 
one of its descendants.

=head2 Deriving a class from one of these classes

If you want to make a new hash class, all I can say is: I<vas-y : fonce !> 
(which means as much as: "go for it!"). There are a few points you have to 
keep in mind, though:

=over 4

=item methods you should override in any case

=over 4

=item read & _write

The read function reads a file into a hash - assuming the file has been 
written by write before. The _write method writes the entire contents of a 
hash to a file. It uses the write_helper_function that can either be passed 
to write() or to set_write_helper_function() to get the size of the data 
associated with a key. The general file format is explained below. If you 
change the way the keys of a hash work, you should write your own read and 
write methods.

=item cmp_keys, empty_key and hash

These functions depend heavily on the type of key you use for the hash: they 
(resp.) compare two keys (and return 0 of they're the same), check whether a 
key is empty or not (and return true if so, false if not) and return a value 
calculated with the key. The latter must always return the same value for the 
same key: it B<must not> be random. Ideally, two very close keys should result 
in two completely different hash values, but a hash function that always 
returns the same value, though extremely bad, would work.

For the hash function I use for strings and generic keys, bigger keys give 
better results, as I use the Adler 32-bit CRC. I can always substitute it with 
another function later :)

=back

=item variables that are important

If you can avoid it at all, don't mess with the mappings variable, or the 
mapping_type typedef: neither of them is guaranteed to work the same way for 
ever. The key_type typedef will always be there and will remain the same, as 
the write_helper_function and filename variables that will remain the same.

=item why not make descendants of IniHash?

The IniHash is a descendant of the NocaseStringHash, which uses 
NocaseStringHash-es as the value for the first (primary) key. Hence, the 
IniHash is a hash of hashes, which is a bit special. If you want, there is no 
problem with making descendants of this hash, but I would strongly advise to 
use NocaseStringHash in stead, as most of it is practically the same, you can 
use the write_helper_function again (which IniHash doesn't use), you're not 
stuck with IniHash' get() and put() functions, etc.

=back

=head2 Hash file format

The hash file format is fairly simple: it starts with a magic number (HASH for 
base hashes, HASI for IntHash-es, HASS for StringHash-es and 
NocaseStringHash-es), a file format version (0100 for the time being), followed
by the key, the size of the associated data and the data. For HASH, the key is 
the key_type structure's contents; for HASS, the key is the size of the string,
followed by the string and a NUL character; for HASI, the key is just an 
integer. In any of these formats, the size of the data is whatever is returned 
by the write_helper_function function.

=head2 IniHash file format

The IniHash file format closely mimics the .ini configuration file format for 
Windows, with a few caveats: 

=over 4

=item 1

Primary keys are set between square brackets, like so: C<[primary key]>. 
Whatever comes after a primary key on the same line is ignored; whatever comes 
before a primary key is ignored as well.

=item 2

Anything after a hash/pound sign C<#> is ignored. Hence, the # sign is 
completely forbidden in the .ini files (this is a FIXME).

=item 3

Anything before a C<=> is assumed to be a secondary key, anything behind it is 
the associated value. Lines that are not primary keys and don't have a C<=> in 
it are ignored

=item 4

Anything with a C<[> in it is assumed to be a primary key

=item 5

Leading and trailing spaces are ignored for both key names and values

=item 6

Secondary key names may be enveloped in parenthesis C<()>

=back

I know this is not perfect, and will gratefully accept patches, but hey, it 
works (for me).


NEWS
$Id: NEWS,v 1.6 2004/05/11 15:59:21 blytkerchan Exp $

New News
========
version 0.2-alpha1
------------------
Mon, May 07, 2004: libcontain-0.2-alpha1 released

This release introduces a number of important new features including a heap and
a map, both based on a lock-free binomial tree. It also includes the required 
memory management to make the containers impervious to the ABA problem and thus
truly thread-safe.

Many of the restrictions formerly bound to the containers have been lifted by 
the integration of the SMR-based memory management, provided by libmemory. 
Most of the containers are now truly thread-safe in the sense that you can 
combine any operation with any other on a single object, including two or more 
write operations.

Additionally, the hash implementation from libhash - with which this library 
began - has been reviewed and rendered thread-safe and faster than it was 
before. This same hash implementation was also the basis of a dictionary object
that outran Microsoft's equivalent. This implementation should now be faster 
than the dictionary in question! It is, however, still lock-based - but a 
non-blocking hash implementation has been on the drawing boards for quite some
time now and should be forthcoming shortly.

Help is still wanted for this project: there is an unsolved race condition in 
the array (vector) implementation and the architecture-specific code needs 
porting to non-IA32 platforms.

version 0.1-alpha1
------------------
Mon, Mar 08, 2004: To be more helpful for the cgdb project, an implementation
of a hashtable written only in C was needed. For the time being, the one from
glib was "ported" into libcontain.
The license of glib is the GNU LGPL, which is strictly speaking not compatible 
with the Jail license the rest of the library falls unde falls under. To not 
have the effects of that license on your program (which will only bother you
if you have a GPL-waiver from me) don't use the glib implementation: configure
with the ``--without-glib'' option.
The reason for using glib was that it's implemented in C whereas the provided
implementation is written in C++. You can turn the provided implementation 
off with ``--without-cxx''.
Note that if you do both --without-cxx and --without-glib, you're left without
a hash table implementation.

version 0.0-alpha3
------------------
Thu, Feb 12, 2004: In my list implementation, the variables marked as "private"
in the algorithm description were put in the list_t type, which basically means
they were part of the list descriptor. That means that they were inherently 
shared among threads, which means that the implementation wasn't thread-safe
at all.
I've created a new type, list_state_t, which is used in the list implementation
by each of the functions that calls list_find(). list_find and all of its
callers will use a local copy of the state, which is allocated on the stack
and which is therefore local to the thread. This should make the list 
thread-safe.

Old News
========
version 0.0-alpha2
------------------
Tue, Feb 10, 2004: I've just finished implementing the lists described by 
M.M. Micheal, though I haven't used his "tag" to prevent the ABA problem:
I will be using the SMR memory management method later on, and I don't have
a CAS that will take both the pointer and the tag at the same time, at the
moment. (That, and I don't want to count on the alignment of the structure
in memory...).
Apart from that, a major new thing is that all of the hash classes actually 
compile now, and that the final library actually contains all of the containers
which was not the case in the previous release.
We still need testcases for all of the non-hash objects, though.., and the
README file isn't quite up-to-date yet..

version 0.0-alpha1
------------------
Tue, Feb 3, 2004: first alpha release of the container library - no testing 
has been done whatsoever, except on the hash. I haven't even taken the time to
re-read the documentation yet. The implementations should work, though, and 
have been tested marginally separately, but as the testsuite hasn't been 
implemented yet, there's not much to guarantee that anything here will
actually work.

Geriatric  News
===============
Oct 5, 2003: integrated in Jail
Aug 15, 2002: First alpha version of the library ready for release.



Hash.cc
/* $Id: Hash.cc,v 1.4 2004/04/07 13:46:27 blytkerchan Exp $ */
/* Jail: Just Another Interpreted Language
 * Copyright (c) 2003, Ronald Landheer-Cieslak
 * All rights reserved
 * 
 * This is free software. You may distribute it and/or modify it and
 * distribute modified forms provided that the following terms are met:
 *
 * * Redistributions of the source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer;
 * * Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in
 *   the documentation and/or other materials provided with the distribution;
 * * None of the names of the authors of this software may be used to endorse
 *   or promote this software, derived software or any distribution of this 
 *   software or any distribution of which this software is part, without 
 *   prior written permission from the authors involved;
 * * Unless you have received a written statement from Ronald Landheer-Cieslak
 *   that says otherwise, the terms of the GNU General Public License, as 
 *   published by the Free Software Foundation, version 2 or (at your option)
 *   any later version, also apply.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
/*** 
// Includes */
#if !defined(_WIN32) || defined (__CYGWIN__)
#include <config.h>
#else  // in native windows
#define HAVE_STRING_H 1
#endif
#include <assert.h>
#include "Hash.h"
#include <stdlib.h>

#ifdef HAVE_STRING_H
	#include <string.h>
#else 	// !HAVE_STRING
	#ifdef HAVE_STRINGS_H
		#include <strings.h>
	#else // !HAVE_STRINGS
		#error Neither string.h nor strings.h is available
		#error Either file is required for the operation of this library
		#error Please contact ronald@landheer.com for more information
	#endif // !HAVE_STRINGS
#endif 	// !HAVE_STRING

/***
// Constants */
static const char *HASH_MAGIC = "HASH";
static const char *HASH_VERSION = "0100";

/*** 
// Local functions */
/** 
// get_prime(): get a prime greater than or equal to _U, return ~0 on error (_U > 3355070839)
*/
ulong get_prime(ulong _u) {
	static const ulong primes[] = {
		19, 29, 41, 59, 79, 107, 149, 197, 263, 347, 457, 599, 787, 1031,
		1361, 1777, 2333, 3037, 3967, 5167, 6719, 8737, 11369, 14783,
		19219, 24989, 32491, 42257, 54941, 71429, 92861, 120721, 156941,
		204047, 265271, 344857, 448321, 582821, 757693, 985003, 1280519,
		1664681, 2164111, 2813353, 3657361, 4754591, 6180989, 8035301,
		10445899, 13579681, 17653589, 22949669, 29834603, 38784989,
		50420551, 65546729, 85210757, 110774011, 144006217, 187208107,
		243370577, 316381771, 411296309, 534685237, 695090819, 903618083,
		1174703521, 1527114613, 1985248999, 2580823717UL, 3355070839UL
	};
	size_t i;
	
	for (i = 0; i < (sizeof(primes) / sizeof(ulong)); i++) {
		if (primes[i] >= _u) return (primes[i]);
	} // for

	return (~0);
} // get_prime()

/*** 
// Constructors */
/** 
// Hash(void): construct a default empty hash
*/
Hash::Hash(void) {
	Count = 0;
	size = get_prime(Count); // FIXME: check this for errors (1)
	mappings = (mapping_type*)malloc(size * sizeof(mapping_type));
	memset(mappings, 0, size * sizeof(mapping_type));
} // Hash()

/** 
// Hash(ulong): construct a default empty hash with initial size C
// NOTE: this is not tested by the test1 test program
*/
Hash::Hash(ulong C) {
	Count = 0;
	size = get_prime(C); // FIXME: check this for errors (1)
	mappings = (mapping_type*)malloc(size * sizeof(mapping_type));
	memset(mappings, 0, size * sizeof(mapping_type));
} // Hash()

/*** 
// Destructors */
/** 
// ~Hash(): free everything allocated
*/
Hash::~Hash(void) {
	if (mappings) free(mappings);
	mappings = NULL;
} // ~Hash()

/*** 
// Public class functions */
/** 
// get(): gets the value corresponding to a key, returns NULL if it doesn't exist
*/
void *Hash::get(const void *key) {
	return(_get(key));
} // get()

void *Hash::_get(const void *key) {
	int location = hash(key) % size;
	ulong i;
	
	for (i = location; (i < size) && !empty_key(mappings[i].key); i++) {
		if (cmp_keys(key, mappings[i].key) == 0) return(mappings[i].value);
	} // for
		
	return (NULL);
} // _get()

/** 
// put(): puts the value VALUE in the hash at key KEY
//        returns true if successful, false if not
*/
bool Hash::put(const void *key, const void *value) {
	return(_put(key, value));
} // put

bool Hash::_put(const void *key, const void *value) {
	int location = hash(key) % size;
	ulong i;
	
	if (((double)(Count + 1) / (double)size) >= 0.75) { // FIXME: make the 0.75 a define - perhaps in config.h?
		if (!grow(size + 1)) return (false);
		return(_put(key, value));
	} // if
	for (i = location; true; i++) {
		if (i == size) {
			if (grow(size + 1)) 
				return (_put(key, value));
			else return (false);
		} // if
		if (empty_key(mappings[i].key)) Count++;
		if (empty_key(mappings[i].key) || (cmp_keys(mappings[i].key, key) == 0)) {
			mappings[i].key = (void*)key;
			mappings[i].value = (void*)value;
			return (true);
		} // if
	} // for

	return(false); // never reached (normally)
} // _put()

/** 
// remove(): remove an entry from the hash
//           we rehash when we've removed the entry to make sure all 
//           entries end up in the right place
// FIXME: this could probably be done more efficiently
*/
bool Hash::remove(const void *key) {
	return(_remove(key));
} // remove()

bool Hash::_remove(const void *key) {
	mapping_type *old_mappings = mappings;
	int location = hash(key) % size;
	ulong i;
	
	for (i = location; (i < size) && !empty_key(old_mappings[i].key); i++) {
		if (cmp_keys(key, old_mappings[i].key) == 0) {
			old_mappings[i].key = NULL;
			mappings = (mapping_type*)malloc(size * sizeof(mapping_type));
			memset(mappings, 0, size * sizeof(mapping_type));
			Count = 0;
			for (i = 0; i < size; i++) {
				if (!empty_key(old_mappings[i].key)) {
					_put(old_mappings[i].key, old_mappings[i].value);
				} // if
			} // for
			free(old_mappings);
			return (true);
		} // if
	} // for
		
	return (false);
} // remove()

/** 
// clear(): clear the hash of all its contents
*/
bool Hash::clear(void) {
	memset(mappings, 0, size * sizeof(mapping_type));
	Count = 0;
		
	return (true);
} // clear()

/** 
// count(): count the number of entries in the hash
*/
ulong Hash::count(void) {
	return (Count);
} // count()

/** 
// keys(): return a list of all keys in the hash
//         returns NULL on error or if there are no keys, otherwise, a pointer to a static array ending with a NULL
// NOTE: this could be done more efficiently (for time) if we cached the list..?
*/
void **Hash::keys(void) {
	return(_keys());
} // keys

void **Hash::_keys(void) {
	static void **keys = NULL;
	unsigned int num_keys = 0;
	ulong i;
	
	if ((keys = (void**)malloc((this->Count + 1) * sizeof(void*))) == NULL) return (NULL);
	for (i = 0; i < size; i++) {
		if (!empty_key(mappings[i].key)) {
			keys[num_keys] = mappings[i].key;
			keys[num_keys + 1] = NULL;
			num_keys++;
		} // if
	} // for
	assert(Count == num_keys);
	
	return(keys);
} // keys()

/** 
// contains(): return true if the key exists, false if not
// NOTE: this could be done more efficiently (for time) if we cashed the list..?
*/
bool Hash::contains(const void *key) {
	return(_contains(key));
} // contains

bool Hash::_contains(const void *key) {
	int location = hash(key) % size;
	ulong i;
	
	for (i = location; i < size && !empty_key(mappings[i].key); i++) {
		if (cmp_keys(key, mappings[i].key) == 0) return true;
	} // for
	
	return false;
} // _contains()

/** 
// cmp_keys(): compare two keys, return 0 if they're the same, nonzero if not
*/
int Hash::cmp_keys(const void *key1, const void *key2) {
	key_type *K1 = (key_type*)key1, *K2 = (key_type*)key2;

	if (K1 == K2) return (0);
	if ((K1 == NULL) || (K2 == NULL)) return (-1);
	if (K1->size != K2->size)
		return (K2->size - K1->size);
	return (memcmp(K1->value, K2->value, K1->size));
} // cmp_keys()

/** 
// hash(): calculates a 32-bit Adler CRC for *VALUE, used to hash the value.
*/
ulong Hash::hash(const void *key) {
	key_type *K = (key_type*)key;
	char *value = (char*)K->value;
	ulong currentAdler1 = 0x00000000UL;
    	ulong currentAdler2 = 0x00000001UL;
    	int i;
    	
    	for (i = 0; i < K->size; i++) {
    		currentAdler1 += value[i];
    		currentAdler2 += currentAdler1;
    	} // for
    	currentAdler1 = currentAdler1 << 16;
    	
    	return (currentAdler1 + currentAdler2);
} // hash()

/** 
// grow(): grow the hash to S (minimum)
*/
bool Hash::grow(int S) {
	mapping_type *old_mappings = mappings;
	ulong i, old_size;

	old_size = size;
	S = get_prime(S);
	if ((mappings = (mapping_type*)malloc(S * sizeof(mapping_type))) == NULL) return (false);
	memset(mappings, 0, S * sizeof(mapping_type));
	Count = 0;
	size = S;
	for (i = 0; i < old_size; i++) {
		if (!empty_key(old_mappings[i].key)) {
			_put(old_mappings[i].key, old_mappings[i].value);
		} // if
	} // for
	free(old_mappings);
	
	return(true);
} // grow()

/** 
// empty_key(): check whether a key is empty - return true if so, false if not
*/
bool Hash::empty_key(void* key) {
	key_type *K = (key_type *)key;
	
	if (K == NULL) return (true);
	return(K->size == 0);
} // empty_key()

bool Hash::read(char *filename) {
	strcpy(this->filename, filename);
	FILE *file;
	int data_size;
	static key_type *key;
	char magic[4];
	void *data = NULL;
	
	assert(sizeof(int) == 4);
	if ((file = fopen(filename, "rb")) == NULL) return(false);
	// check the magic number
	fread(magic, 4, 1, file);
	if (memcmp(magic, HASH_MAGIC, 4) != 0) return(false);
	fread(magic, 4, 1, file);
	if (memcmp(magic, HASH_VERSION, 4) != 0) return(false);
	while(fread(&data_size, 4, 1, file) == 1) {
		key = (key_type*)malloc(sizeof(key_type));
		key->value = NULL; // just to make sure
		key->size = data_size;
		key->value = malloc(key->size);
		fread(key->value, key->size, 1, file);
		fread(&data_size, 4, 1, file);
		data = malloc(data_size);
		fread(data, data_size, 1, file);
		this->_put(key, data);
	} // while
	fclose(file);
	
	return(true);
} // read()

bool Hash::write(char *filename, int (*helper_function)(void*)) {
	strcpy(this->filename, filename);
	write_helper_function = helper_function;
	return(this->write());
} // write()

bool Hash::write(int (*helper_function)(void*)) {
	write_helper_function = helper_function;
	return(this->write());
} // write()

bool Hash::write(char *filename) {
	strcpy(this->filename, filename);
	return(this->write());
} // write()

bool Hash::write(void) {
	return(_write());
} // write

bool Hash::_write(void) {
	key_type **key_list;
	int i, data_size;
	FILE *file;
	void *data;
	
	assert(sizeof(int) == 4);
	if ((file = fopen(filename, "wb")) == NULL) return(false);
	fwrite(HASH_MAGIC, 4, 1, file);
	fwrite(HASH_VERSION, 4, 1, file);
	key_list = (key_type**)this->keys();
	for (i = 0; key_list[i]; i++) {
		fwrite(&key_list[i]->size, 4, 1, file);
		fwrite(key_list[i]->value, key_list[i]->size, 1, file);
		data = this->_get(key_list[i]);
		data_size = write_helper_function(data);
		fwrite(&data_size, 4, 1, file);
		fwrite(data, data_size, 1, file);
	} // for
	fclose(file);
	
	return(true);
} // write()

void Hash::set_write_helper_function(int (*helper_function)(void*)) {
	write_helper_function = helper_function;
} // set_write_helper_function()

void * Hash::get_write_helper_function(void) {
	return((void*)write_helper_function);
} // set_write_helper_function()

char *Hash::get_filename(void) {
	return(filename);
} // get_filename()

void Hash::set_filename(char *filename) {
	strcpy(this->filename, filename);
} // get_filename()

// rlc	2004-03-31
// added to implement hash_search
void Hash::for_each(foreach_func_t func, void * user_data)
{
	ulong i;
	
	for (i = 0; i < Count; i++)
	{
		if (mappings[i].key)
			func(mappings[i].key, mappings[i].value, user_data);
	}
}

/* FOOTNOTES
// 1. Constructors do not return any value, so we should get back to the good old RunControl variable..
*/

